pkg/planner/core/indexmerge_path.go

```diff
@@ -478,6 +595,159 @@ func (ds *DataSource) generateIndexMergeAndPaths(normalPathCnt int) *util.Access
+			if err != nil {
+				logutil.BgLogger().Debug("build index merge partial mv index paths failed", zap.Error(err))
+				return nil, nil, false, err
+			}
+			if isIntersection || !ok { // limitation 2
+				continue
+			}
+			if len(paths) == 0 {
+				continue
+			}
+			// UNION case, use the max count after access for simplicity.
+			maxCountAfterAccess := -1.0
+			for _, p := range paths {
+				maxCountAfterAccess = math.Max(maxCountAfterAccess, p.CountAfterAccess)
+			}
+			// Note that: here every path is about mv index path.
+			// find the most valuable mv index path, which means it has the minimum countAfterAccess.
+			if len(bestPaths) == 0 {
+				bestPaths = paths
+				bestCountAfterAccess = maxCountAfterAccess
+				bestNeedSelection = len(remainingFilters) != 0
+			} else {
+				if bestCountAfterAccess > maxCountAfterAccess {
+					bestPaths = paths
+					bestCountAfterAccess = maxCountAfterAccess
+					bestNeedSelection = len(remainingFilters) != 0
+				}
+			}
+		}
+		if len(bestPaths) != 0 {
+			usedMap[offset] = true
+			// correctly find a dnf condition for this mv index path
+			mvAndPartialPaths = append(mvAndPartialPaths, bestPaths...)
+			if !needSelection && bestNeedSelection {
+				// collect one path's need selection flag.
+				needSelection = bestNeedSelection
+			}
+		}
+	}
+	return mvAndPartialPaths, usedMap, needSelection, nil
+}
+
+func (ds *DataSource) generateIndexMergePartialPaths4And(normalPathCnt int,
+	indexMergeConds []expression.Expression) ([]*util.AccessPath, map[string]expression.Expression, error) {
+	// step1: collect all mv index paths
+	possibleMVIndexPaths := make([]*util.AccessPath, 0, len(ds.possibleAccessPaths))
+	for idx := 0; idx < normalPathCnt; idx++ {
+		if !isMVIndexPath(ds.possibleAccessPaths[idx]) {
+			continue // not a MVIndex path
+		}
+		possibleMVIndexPaths = append(possibleMVIndexPaths, ds.possibleAccessPaths[idx])
+	}
+	// step2: mapping index merge conditions into possible mv index path
+	mvAndPartialPath := make([]*util.AccessPath, 0, len(possibleMVIndexPaths))
+	usedAccessCondsMap := make(map[string]expression.Expression, len(indexMergeConds))
+	for idx := 0; idx < len(possibleMVIndexPaths); idx++ {
+		idxCols, ok := ds.prepareCols4MVIndex(possibleMVIndexPaths[idx].Index)
+		if !ok {
+			continue
+		}
+		accessFilters, _, mvColOffset, mvFilterMutations := ds.collectFilters4MVIndexMutations(indexMergeConds, idxCols)
+		if len(accessFilters) == 0 { // cannot use any filter on this MVIndex
+			continue
+		}
+		// in traveling of these mv index conditions, we can only use one of them to build index merge path, just fetch it out first.
+		// build index merge partial path for every mutation combination access filters.
+		for _, mvFilterMu := range mvFilterMutations {
+			// derive each mutation access filters
+			accessFilters[mvColOffset] = mvFilterMu
+
+			partialPaths, isIntersection, ok, err := ds.buildPartialPaths4MVIndex(accessFilters, idxCols, possibleMVIndexPaths[idx].Index)
```

Comment:
Should we copy the whole `accessFilters` here? I'm not sure whether it will be overwritten if we don't copy it.

Reply:
you are right, make senseeeee.